src/qibo/tensorflow/custom_operators/cc/kernels/measurements.cc

```diff
@@ -0,0 +1,121 @@
+#if GOOGLE_CUDA
+#define EIGEN_USE_GPU
+#endif  // GOOGLE_CUDA
+
+#include "measurements.h"
+#include "tensorflow/core/framework/op_kernel.h"
+
+namespace tensorflow {
+
+typedef Eigen::ThreadPoolDevice CPUDevice;
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace functor {
+
+// CPU specialization
+template <typename Tint, typename Tfloat>
+struct MeasureFrequenciesFunctor<CPUDevice, Tint, Tfloat> {
+  void operator()(const CPUDevice &d, Tint* frequencies, const Tfloat* probs,
+                  Tint nshots, int nqubits, int user_seed = 1234)
+  {
+    int64 nstates = 1 << nqubits;
+    srand(user_seed);
+    // Create vector of seeds for each thread
+    std::vector<unsigned> thread_seed;
+    for (auto i = 0; i < omp_get_max_threads(); i++) {
+      thread_seed.push_back(rand());
+    }
+    // Initial bitstring is the one with the maximum probability
+    int64 initial_shot = 0;
+    for (int64 i = 0; i < nstates; i++) {
+        if (probs[i] > probs[initial_shot]) {
+            initial_shot = i;
+        }
+    }
+    #pragma omp parallel
+    {
+        std::unordered_map<int64, int64> frequencies_private;
+        unsigned seed = thread_seed[omp_get_thread_num()];
+        int64 shot = initial_shot;
+        #pragma omp for
+        for (int64 i = 0; i < nshots; i++) {
+          // Generate random index to flip its bit
+          int flip_index = ((int) rand_r(&seed) % nqubits);
+          // Flip the corresponding bit
+          int current_value = ((int64) shot >> flip_index) % 2;
+          int64 new_shot = shot + ((int64)(1 - 2 * current_value)) * ((int64) 1 << flip_index);
+          // Accept or reject move
+          Tfloat ratio = probs[new_shot] / probs[shot];
+          if (ratio > ((Tfloat) rand_r(&seed) / RAND_MAX)) {
+            shot = new_shot;
+          }
+          // Update frequencies
+          if (frequencies_private.find(shot) == frequencies_private.end()) {
+              frequencies_private[shot] = 1;
+          } else {
+              frequencies_private[shot]++;
+          }
+        }
+        #pragma omp critical
+        {
+            for(const auto& entry : frequencies_private) {
+                frequencies[entry.first] += entry.second;
+            }
+        }
+    }
+  }
+};
+
+template <typename Device, typename Tint, typename Tfloat>
+class MeasureFrequenciesOp : public OpKernel {
+ public:
+  explicit MeasureFrequenciesOp(OpKernelConstruction *context) : OpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("nqubits", &nqubits_));
+    OP_REQUIRES_OK(context, context->GetAttr("omp_num_threads", &threads_));
+    OP_REQUIRES_OK(context, context->GetAttr("seed", &seed_));
+    omp_set_num_threads(threads_);
+  }
+
+  void Compute(OpKernelContext *context) override {
+    // grab the input tensor
+    Tensor frequencies = context->input(0);
+    const Tensor& cumprobs = context->input(1);
+    const Tensor& nshots = context->input(2);
```

Comment:
@stavros11, for the GPU would be great if we could consider `nshots` as attribute, otherwise we cannot access its value for the GPU block/thread computation.

Reply:
The issue is that if we define `nshots` as attribute:
```C++
REGISTER_OP("MeasureFrequencies")
    .Attr("Tfloat: {float32, float64}")
    .Attr("Tint: {int32, int64}")
    .Input("frequencies: Tint")
    .Input("probs: Tfloat")
    .Attr("nshots: int")
    ...
```
then its type is `int32` and this creates problems for many shots (like 10^10). I am not sure if Tensorflow supports int64 for attributes, but I think no according to their [custom op guide](https://www.tensorflow.org/guide/create_op#attr_types).

Reply:
Right, so I think a potential solution is to store the `nshots` value as the shape of the input tensor, so we can access this information from CPU. What do you think?

Reply:
There is no tensor with shape that involves `nshots`, do you mean creating a new tensor of that shape as input? Won't this cause memory problems when nshots is very large?

Reply:
Yes, this will cause memory issues. What about always `float/double` and then we cast to int64?

Reply:
I think attributes support only float so this would have the same issue as `int32`, right?

Reply:
In terms of range, float should work, the range is between 1.2E-38 to 3.4E+38.

Maybe a more elegant solution is to always copy the `nshots` to CPU by allocating a new variable using `with tf.device('CPU'):` and then calling the measurement operator. This should allow us to keep using the int64 but also access kernels from custom operators.

Reply:
> Maybe a more elegant solution is to always copy the `nshots` to CPU by allocating a new variable using `with tf.device('CPU'):` and then calling the measurement operator. This should allow us to keep using the int64 but also access kernels from custom operators.

Would this require any change in the custom operator or we leave nshots as it is (input) and we just cast it on CPU in the Python code? Would there be any performance issues with the GPU due to copying nshots from CPU?

Reply:
Yes, we keep the op unchanged but overload the python method with the cpu cast. I don't think we will see memory issues, if I understand the code, nshots will be always on a cpu tensor. Cod you please give a try?

Reply:
I am not sure if I understood correctly how this will help in the CPU case but can you please check if what I pushed in the last commit is correct? The reason I did not overload the python method in `qibo_tf_custom_operators.py` is that I don't want to import K there (it leads to a circular import so we would have to import it from within the function). 

The `measure_frequencies` op is only used by `sample_frequencies` in the Tensorflow backend so what I did should be equivalent to overloading the method.

Reply:
Thanks. I have tested, however nshots is copied back to the GPU after the cast. So this approach doesn't work.